source/extensions/quic_listeners/quiche/quic_filter_manager_connection_impl.cc

```diff
@@ -50,20 +50,43 @@ bool QuicFilterManagerConnectionImpl::aboveHighWatermark() const {
 }
 
 void QuicFilterManagerConnectionImpl::close(Network::ConnectionCloseType type) {
-  if (type != Network::ConnectionCloseType::NoFlush) {
-    // TODO(danzh): Implement FlushWrite and FlushWriteAndDelay mode.
-  }
   if (quic_connection_ == nullptr) {
     // Already detached from quic connection.
     return;
   }
-  closeConnectionImmediately();
-}
-
-void QuicFilterManagerConnectionImpl::setDelayedCloseTimeout(std::chrono::milliseconds timeout) {
-  if (timeout != std::chrono::milliseconds::zero()) {
-    // TODO(danzh) support delayed close of connection.
-    NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+  const bool delayed_close_timeout_configured = delayed_close_timeout_.count() > 0;
+  if (hasDataToWrite() && type != Network::ConnectionCloseType::NoFlush) {
+    // QUIC connection has unsent data and caller wants to flush them. Wait for flushing or timeout.
+    if (!inDelayedClose() && delayed_close_timeout_configured) {
+      // Only set alarm if not in delay close mode yet.
+      initializeDelayedCloseTimer();
+    }
+    // Update delay close state according to current call.
+    if (delayed_close_timeout_configured &&
+        type == Network::ConnectionCloseType::FlushWriteAndDelay) {
+      delayed_close_state_ = DelayedCloseState::CloseAfterFlushAndWait;
+    } else {
+      delayed_close_state_ = DelayedCloseState::CloseAfterFlush;
+    }
+  } else {
```

Comment:
I think we could remove this else, and just have else if (hasDataToWrite()) {} else { // type == NoFlush